docs/src/proposals/secp256k1_instruction.md

```diff
@@ -0,0 +1,75 @@
+data:
+
+```
+struct SignatureOffsets {
+    signature_offset: u16,        // offset to [signature,recovery_id] of 64+1 bytes
+    signature_account_index: u8,  // account index of signature data
+    pubkey_offset: u16,           // offset to pubkey in ethereum format of keccak256(pubkey[1..])[12..]
+    pubkey_account_index: u8,     // index of account to receive pubkey
+    message_data_offset: u16,     // offset to start of message data
+    message_data_size: u16,       // size of message data
+    message_account_index: u8,    // index of account to receive message data
+}
+```
+
+Pseudo code of the operation:
+```
+for i in 0..count {
+    // i'th index values referenced:
+    signature = accounts[signature_account_index].data[signature_offset..signature_offset + 64]
+    recovery_id = accounts[signature_account_index].data[signature_offset + 64]
+    message_hash = keccak256(accounts[message_account_index].data[message_data_offset..message_data_offset + message_data_size])
+    pubkey = ecrecover(signature, recovery_id, message_hash)
+    eth_pubkey = keccak256(pubkey[1..])[12..]
+    if eth_pubkey != accounts[pubkey_account_index].data[pubkey_offset..pubkey_offset + 20] {
+       return Error
+    }
+}
+return Success
+```
+
+This allows the user to specify the source account and offset for any of the signature, pubkey, and messaage data.
+By specifying a special instructions sysvar, one can also receive data from the transaction itself.
+
+Cost of the transaction will count the number of signatures to verify multiplied by the signature cost verify multiplier.
+
+## Optimization notes
+
+The operation will have to take place after deserialization and accounts load assuming the transaction references external
+accounts for the inputs. It should be relatively easy though to scan the transaction for introspection inputs.
+
+## Other solutions
+
+* Instruction available as CPI such that the program can call as desired or a syscall which can operate on the instruction inline.
+   - Could be harder to optimize given that it generally either requires bpf program scan to determine the inputs to the operation,
+     or the implementation needs to just wait until the program hits the operation in bpf processing to evaluate it.
+   - Vector version of the operation could allow for somewhat efficient simd/gpu execution. For most efficient though,
+     batching with other instructions in the pipeline would be ideal.
+   - Pros - Nicer interface for the user.
+
+* Async execution environment
+   - Might be hard to optimize for devices like gpus which cannot queue work for itself easily
+   - Might be easier to optimize on cpu since ordering can be more explicit
+
+* All inputs have to come from the instruction
+   - Pros - easier to optimize, data is already sent to the GPU for instance for regular sigverify. Probably still need to
```

Comment:
This looks like the easiest way to start. You could always add additional instructions later to reduce transaction space, if needed.